### PR TITLE
#32458 fix for of loop in internet explorer

### LINKED ISSUE #32458: for of does not work on generator function when running on es5 aka internet explorer
**TypeScript Version:**  3.5.3** 

** Related issue **
https://github.com/microsoft/tslib/issues/59

downleveliteration es5, downleveliteration internet explorer, generator internet explorer

**Code**

```ts
function* issue()
{
    yield "hello";
    yield "world";
}

for (let x of issue())
{
    console.log(x);
}

**Expected behavior:**
hello, world is logged to the console when running on internet explorer 11

**Actual behavior:**
console logs undefined forever as stuck in loop

** Fix **
The code in tslib / which is added to top of transpiled file should check if we are enumerating over an array or have something which is IterableIterator like

``` js
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (m) return m.call(o);
    if (Array.isArray(o)) {
        var i = 0;
        return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
    } else if (typeof o.next === "function") {  // should probably be more concise
        return o;
    }  else {
    // kittens die
    }
};

** Relatd
@microsoft/tslib#59

### PR DESCRIPTION
**Pull request by Saulzi**
Fixes #32458

### CODE REVIEW

File: src/testRunner/unittests/evaluation/forOf.ts [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        const result = evaluator.evaluateTypeScript(`
+        export function main() {
+            let x = undefined;
+
+            for (let value of x) {
+            }
+        }
+        `, { downlevelIteration: true, target: ts.ScriptTarget.ES5 });
+
+        assert.throws(() => result.main(), "undefined is not iterable (cannot read property Symbol(Symbol.iterator))");
```
Comment (rbuckton) [inline src/testRunner/unittests/evaluation/forOf.ts]:
```suggestion
        assert.throws(() => result.main(), /cannot read property.*Symbol\(Symbol\.iterator\).*/i);
```

File: src/compiler/factory.ts [hunk 2]
```diff
@@ -3636,15 +3636,16 @@ namespace ts {
...
+                var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
                 if (m) return m.call(o);
-                return {
+                if (o && typeof o.length === "number") return {
                     next: function () {
                         if (o && i >= o.length) o = void 0;
                         return { value: o && o[i++], done: !o };
                     }
                 };
+                throw new TypeError(s ? "Object not iterable." : "Symbol.iterator is not defined.");
```
Comment (DanielRosenwasser) [inline src/compiler/factory.ts]: Can you make this "Object **is** not iterable."?

Comment (rbuckton) [reply src/compiler/factory.ts]: Adding a suggestion.

```suggestion
                throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
```

**Comment by rbuckton** [comment]
@Saulzi: I'm going to commit the two suggestions to this PR and once the build succeeds I'll go ahead and merge.

File: src/testRunner/unittests/evaluation/forOf.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        const result = evaluator.evaluateTypeScript(`
+        export function main() {
+            let x = {} as any;
+
+            for (let value of x) {
+            }
+        }
+        `, { downlevelIteration: true, target: ts.ScriptTarget.ES5 });
+
+        assert.throws(() => result.main(), "Object not iterable");
```
Comment (rbuckton) [inline src/testRunner/unittests/evaluation/forOf.ts]:
```suggestion
        assert.throws(() => result.main(), "Object is not iterable");
```

File: tests/baselines/reference/ES5For-of33.js [hunk 4]
```diff
@@ -4,15 +4,16 @@ for (var v of ['a', 'b', 'c']) {
...
+    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
     if (m) return m.call(o);
-    return {
+    if (o && typeof o.length === "number") return {
         next: function () {
             if (o && i >= o.length) o = void 0;
             return { value: o && o[i++], done: !o };
         }
     };
+    throw new TypeError(s ? "Object not iterable." : "Symbol.iterator is not defined.");
```
Comment (rbuckton) [inline tests/baselines/reference/ES5For-of33.js]:
```suggestion
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
```

File: tests/baselines/reference/ES5For-of33.sourcemap.txt [hunk 5]
```diff
@@ -8,15 +8,16 @@ sources: ES5For-of33.ts
...
+>>>    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
 >>>    if (m) return m.call(o);
->>>    return {
+>>>    if (o && typeof o.length === "number") return {
 >>>        next: function () {
 >>>            if (o && i >= o.length) o = void 0;
 >>>            return { value: o && o[i++], done: !o };
 >>>        }
 >>>    };
+>>>    throw new TypeError(s ? "Object not iterable." : "Symbol.iterator is not defined.");
```
Comment (rbuckton) [inline tests/baselines/reference/ES5For-of33.sourcemap.txt]:
```suggestion
>>>    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
```

File: src/compiler/factory.ts (hunk 2 above)
Comment (rbuckton) [reply src/compiler/factory.ts]: There are too many tests to resolve doing these via suggestions...

**Comment by rbuckton** [comment]
It looks like I'm unable to update the branch to commit these suggestions. @Saulzi are you able to take a look at this?

**Comment by rbuckton** [comment]
It looks like I can merge manually and make the minor fixes when I merge.

**Comment by rbuckton** [comment]
Thanks for the contribution!

**Pull request merged**